tests/gtest/avif_fuzztest_enc_dec_incr_experimental.cc

```diff
@@ -0,0 +1,127 @@
+// Copyright 2023 Google LLC
+// SPDX-License-Identifier: BSD-2-Clause
+// Encode a fuzzed image split into a grid and decode it incrementally.
+// Compare the output with a non-incremental decode.
+
+#include <cassert>
+#include <cstdint>
+#include <vector>
+
+#include "avif/internal.h"
+#include "avif_fuzztest_helpers.h"
+#include "avifincrtest_helpers.h"
+#include "aviftest_helpers.h"
+#include "fuzztest/fuzztest.h"
+#include "gtest/gtest.h"
+
+using ::fuzztest::Arbitrary;
+using ::fuzztest::InRange;
+
+namespace libavif {
+namespace testutil {
+namespace {
+
+::testing::Environment* const stack_limit_env =
+    ::testing::AddGlobalTestEnvironment(
+        new FuzztestStackLimitEnvironment("524288"));  // 512 * 1024
+
+// Encodes an image into an AVIF grid then decodes it.
+void EncodeDecodeGridValid(AvifImagePtr image, AvifEncoderPtr encoder,
+                           AvifDecoderPtr decoder, uint32_t grid_cols,
+                           uint32_t grid_rows, bool is_encoded_data_persistent,
+                           bool give_size_hint_to_decoder) {
+  ASSERT_NE(image, nullptr);
+  ASSERT_NE(encoder, nullptr);
+
+  const std::vector<AvifImagePtr> cells =
+      ImageToGrid(image.get(), grid_cols, grid_rows);
+  if (cells.empty()) return;
+  const uint32_t cell_width = cells.front()->width;
+  const uint32_t cell_height = cells.front()->height;
+  const uint32_t encoded_width = std::min(image->width, grid_cols * cell_width);
+  const uint32_t encoded_height =
+      std::min(image->height, grid_rows * cell_height);
+
+  const avifImage* gain_map = image->gainMap.image;
+  if (gain_map != nullptr) {
+    std::vector<AvifImagePtr> gain_map_cells =
+        ImageToGrid(gain_map, grid_cols, grid_rows);
+    if (gain_map_cells.empty()) return;
+    ASSERT_EQ(gain_map_cells.size(), cells.size());
+    for (size_t i = 0; i < gain_map_cells.size(); ++i) {
+      cells[i]->gainMap.image = gain_map_cells[i].release();
```

Comment:
What happens if the gain map has different dimensions than the image? Are the cells still valid?  
This is mostly by curiosity, for fuzzing it does not matter.

Reply:
Thanks for the explanation. Note that the current change does not test mismatched `grid_cols` or `grid_rows` and `colorGrid+gainMapNoGrid` or `colorNoGrid+gainMapGrid` but that is fine.